Escape asterisk characters in group names

esingletary opened the pull request:
In our organization's Active Directory instance, we have group names with asterisks in them (so they show up at the top of the group listing). The function `isUserMemberOf` uses regex to test if  common and distinguished names are valid, which is fine until an asterisk is in the name. It causes an exception because of the unescaped `*` in the regex. This adds a second replace statement that escapes it so when it's used in the RegExp statement, it still evaluates correctly. This fix worked fine for me but if you have a better way to go about it, I'm fine with it as long as it works.

jsumners commented:
Does it not work if you escape the character in your supplied search term?

esingletary commented:
You can escape it in the search term but when it retrieves the list of groups from AD, it doesn't escape asterisks in the names of the groups it's checking against.

For example, my code is checking if the specified user is in a group called `Password Managers_main` and it's giving this [error](https://i.imgur.com/v80Z0w9.png) when I check using a particular user that is a part of a group that contains an asterisk. It's getting caught on a group called `*dl-its-tsc`.

jsumners commented:
What do you think about changing https://github.com/esingletary/node-activedirectory/blob/7e83ead890af1b54ef7fbb2fa8fc036a663206d1/lib/activedirectory.js#L621-L627

To:

```js
const lowerCaseGroupName = _groupName.toLowerCase()
const result = groups.filter((g) => g.dn.toLowerCase() === lowerCaseGroupName || g.cn.toLowerCase().includes(lowerCaseGroupName)).length > 0
```

esingletary commented:
So I changed the code in my local copy and it spat out an error, saying that `dn` was undefined. I changed the code to say:

````
const result = groups.filter((g) => (g.dn || '').toLowerCase() === lowerCaseGroupName || (g.cn || '').toLowerCase().includes(lowerCaseGroupName)).length > 0
````
Which solved the issue, but then I realized when trying a distinguished name in the query that the groups array doesn't actually contain distinguished names, so the first condition in the filter will evaluate to `''` every time and not match.

I realize with the original code the regex would simply get the common name from `lowerCaseGroupName` if the user entered a distinguished name instead of a common name, so it wasn't an issue whether the groups array had distinguished names in them or not, as long as they had common names.

The following is an example of output, where the first line is the group query (a distinguished name) and the result is from a user that we're checking (I obfuscated our domain with contoso.com but otherwise it's exact output):

````
cn=*dl-its-tsc,ou=groups,ou=people,dc=ad,dc=contoso,dc=com
[ Group { cn: 'Everyone_main' },
  Group { cn: 'FACSTAFF_main' },
  Group { cn: 'FACSTAFF_mvnc' },
  Group { cn: 'Login Disabled_misc' },
  Group { cn: '*DL-ITS-TSC' },
  Group { cn: 'colleagueusers' },
  Group { cn: 'busobj_users' },
  Group { cn: 'Staff' },
  Group { cn: 'generic-email-accounts' } ]
````
(I'm making the assumption that not returning distinguished names isn't an issue with our configuration of AD but something with the module itself, I could be wrong.)

Notice that it just fetches common names from Active Directory. One solution would be to just continue to take the common name at the beginning of a distinguished name or to figure out why it's not returning distinguished names in the first place. 

What do you think?

jsumners commented:
I think the problem with the `dn` not being returned might be https://github.com/esingletary/node-activedirectory/blob/7e83ead890af1b54ef7fbb2fa8fc036a663206d1/lib/activedirectory.js#L609 😆

esingletary commented:
Oh wow I didn't notice that hahahah. Yep that did it.

jsumners commented:
Excellent. Would you mind updating your PR accordingly?

jsumners commented:
Damn! Looks like some tests are failing.

esingletary commented:
Hold on, I'll check those out.

esingletary commented:
The tests were failing because of some spaces, which could occur so removing them in all dn's and cn's sorts it out.

jsumners commented:
Awesome. Thank you for putting the time in on this. I will publish a release when I have a few minutes to do so.

jsumners commented:
Published as v1.2.4

esingletary commented:
Thanks! Glad we could fix it :)
